File: ApemanPlugin/src/main/kotlin/apeman_core/features_extraction/calculators/method/NumUsedTypesMethodCalculator.java
package apeman_core.features_extraction.calculators.method;

import apeman_core.base_entities.FeatureType;
import apeman_core.pipes.CandidateWithFeatures;
import apeman_core.utils.MethodUtils;
import apeman_core.utils.TypeUtils;
import com.intellij.psi.*;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.Set;

public class NumUsedTypesMethodCalculator extends MethodCalculator {

    public NumUsedTypesMethodCalculator(ArrayList<CandidateWithFeatures> candidates) {
        super(candidates, FeatureType.CON_TYPE_ACCESS);
    }

    private int methodNestingDepth = 0;
    private Set<PsiType> typeSet = null;

    @Override
    public JavaRecursiveElementVisitor createVisitor() {
        return new Visitor();
    }

    private class Visitor extends JavaRecursiveElementVisitor {

        @Override
        public void visitMethod(PsiMethod method) {
            if (methodNestingDepth == 0) {
                typeSet = new HashSet<PsiType>();
            }
            TypeUtils.addTypesFromMethodTo(typeSet, method);

            methodNestingDepth++;
            super.visitMethod(method);
            methodNestingDepth--;
            if (methodNestingDepth == 0 && !MethodUtils.isAbstract(method)) {
//                postMetric(method, typeSet.size());
            }
        }

        @Override
        public void visitElement(PsiElement element) {
            super.visitElement(element);
            TypeUtils.tryAddTypeOfElementTo(typeSet, element);
        }
    }
}

File: ApemanPlugin/src/main/kotlin/apeman_core/features_extraction/calculators/method/NumUsedTypesMethodCalculator.kt
package apeman_core.features_extraction.calculators.method

import apeman_core.base_entities.FeatureType
import apeman_core.features_extraction.calculators.BaseMetricsCalculator
import apeman_core.pipes.CandidateWithFeatures
import apeman_core.utils.CandidateUtils
import apeman_core.utils.MethodUtils
import apeman_core.utils.TypeUtils
import com.intellij.psi.*

import java.util.ArrayList
import java.util.HashSet

class NumUsedTypesMethodCalculator(candidates: ArrayList<CandidateWithFeatures>
) : BaseMetricsCalculator(candidates, FeatureType.CON_TYPE_ACCESS) {

    private var methodNestingDepth = 0
    private var typeSet: MutableSet<PsiType>? = null

    override fun createVisitor(): JavaRecursiveElementVisitor {
        return Visitor()
    }

    private inner class Visitor : JavaRecursiveElementVisitor() {

        override fun visitMethod(method: PsiMethod) {
            if (methodNestingDepth == 0) {
                typeSet = HashSet()
            }
            TypeUtils.addTypesFromMethodTo(typeSet!!, method)

            methodNestingDepth++
            super.visitMethod(method)
            methodNestingDepth--
            if (methodNestingDepth == 0 && !MethodUtils.isAbstract(method)) {
                val res = typeSet!!.count().toDouble()
                CandidateUtils
                        .getCandidatesOfMethod(method, candidates)
                        .forEach { results.set(it, firstFeature, res) }
            }
        }

        override fun visitElement(element: PsiElement) {
            super.visitElement(element)
            TypeUtils.tryAddTypeOfElementTo(typeSet!!, element)
        }
    }
}
